MudToggleGroup: Reorder Selection Logic for ToggleGroupItems

versile2 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
Previously ToggleGroup repeatedly set the selected state of an item except in cases where the item value did not change. Now it only sets selected states after values are set, separating the logic of value setting and select setting. There will be similar fixes applied to the other groups (Chipset for sure, I will checked others)
resolves #10535  

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Added some unit tests

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

versile2 committed:
move the selection portion of togglegroup after value set, added viewer test

versile2 committed:
unit tests

versile2 committed:
remove references to Value/Values and use _value/_values instead. Also added a null check for when the parameters aren't used to Set the SelectedItem.

ScarletKuro commented on src/MudBlazor/Components/Toggle/MudToggleGroup.razor.cs:
```diff
@@ -333,38 +314,40 @@ protected internal async Task ToggleItemAsync(MudToggleItem<T> item)
-                    item.SetSelected(true);
                     await _value.SetValueAsync(itemValue);
                 }
+                if (Value is null)
```
Shouldn't this be `_value.Value`?

versile2 replied:
No this is where I'm doing something if the user didn't bind or use Value. If they did it fires OnValu Changed.

henon replied:
@ScarletKuro has the right of it here. By introducing ParameterState we introduced a general rule that must always be followed: "From inside the component never directly check the value of a parameter and never set it directly." All interaction with a parameter must happen via the ParameterState object.

ScarletKuro commented on src/MudBlazor/Components/Toggle/MudToggleGroup.razor.cs:
```diff
@@ -333,38 +314,40 @@ protected internal async Task ToggleItemAsync(MudToggleItem<T> item)
-                DeselectAllItems();
-                item.SetSelected(true);
                 await _value.SetValueAsync(itemValue);
+                if (Value is null)
```
Shouldn't this be `_value.Value`?

versile2 replied:
Same as above

ScarletKuro commented:
https://github.com/MudBlazor/MudBlazor/blob/b00378ff57affb6ceeed1643e518352b99b0d379/src/MudBlazor/Components/Toggle/MudToggleGroup.razor.cs#L324
And this `_values.Value`?

ScarletKuro commented:
Adding @henon as he was tightly working with this component.
Apart of my questions, it looks good to me

ScarletKuro commented:
I will reply here instead under the review since the text is long 

> No this is where I'm doing something if the user didn't bind or use Value. If they did it fires OnValu Changed.

Ah, I understand now. Unfortunately, this will break the following scenario:
```razor
<MudToggleGroup T="string" SelectionMode="SelectionMode.SingleSelection" Color="Color.Primary" Value="@("No")" CheckMark FixedContent>
	<MudToggleItem Value="@("Yes")" Text="Yes" />
	<MudToggleItem Value="@("No")" Text="No" />
	<MudToggleItem Value="@("Don't know")" Text="Don't know (Click me twice)" />
</MudToggleGroup>
```
If we set the initial value of `Value` without using two-way binding (`@bind-Value`), the toggle will not be able to change its state and will remain stuck at the `No` value.

Usually, we consider this an initial value (like in `MudCollapse`), not a constant, because if people set `Expanded` to `true`, they would never be able to collapse it. This is why in `ParameterState`, the inner value and the outer parameter value are independent of each other.

For example, if you look at Blazor native components like `InputCheckbox`:
```
<InputCheckbox Value="@val" ValueExpression="() => val" />
@code {
    private bool val = false;
}
```
(We need the `ValueExpression` for `InputCheckbox`, unfortunately, but imagine it's just `InputCheckbox Value="@false"`)

You will see that the component changes its state and doesn't stay at the constant `false`, so I believe our behavior is correct.

I'm not sure if `MudToggleGroup` should be an exception to this rule, but I’ll leave it up to @henon to decide.

versile2 commented:
Yup I should have left it value changed.hasdelegate instead. Correct?

henon commented:
> Yup I should have left it value changed.hasdelegate instead. Correct?

Yes, if you want to do something in case it is not two-way bound then this would be the correct thing to do. Why does it matter if the user did two-way bind or not?

henon commented:
Actually, I don't think the reported issue #10535 is actually a bug per se. We have seen such issues before and they can be worked around really easily by just setting the value, waiting a millisecond and reverting it. like shown in this trymud: https://try.mudblazor.com/snippet/GYcpYlaFIEgHtitF

Not sure if we should compromise our value change detection for such fringe use-cases.

ScarletKuro commented:
> Why does it matter if the user did two-way bind or not?

I guess the logic is that if two-way binding is used, like as `bind-Value` or `Value` + `ValueChanged`, then we should allow intercepting the value like this:

```csharp
private void OnValueChanged(string value)
{
    if (value == "Don't know")
    {
        _value1 = "Yes";
    }
    else
    {
        _value1 = value;
    }
}
```

Currently, with `ParameterState`, we do not allow modifying the two-way value (intercepting it) in the `OnValueChanged` method. The only scenario we allow is setting the value as-is:
```
private void OnValueChanged(string value)
{
    _value1 = value;
    //Custom logic that doesn't change the `_value`, maybe a validation or something else
}
```
And then overriding it with `bind-Value:after`. That works.

If I wanted to support interception in `ParameterState`, this is what I would do as well (check if there is two-way used), and I even have code that works in 99% of scenarios:

```diff
public void OnParametersSet()
{
    var currentParameterValue = _getParameterValueFunc();
    if (!_comparer.Equals(_lastValue, currentParameterValue))
    {
        _isChildOriginatedChange = _comparer.Equals(_value, currentParameterValue);
        _value = currentParameterValue;
        _lastValue = currentParameterValue;
    }
+    // intercept logic
+    if (!_comparer.Equals(_value, currentParameterValue))
+    {
+        var eventCallback = _eventCallbackFunc();
+        if (eventCallback.HasDelegate)
+        {
+            _value = currentParameterValue;
+        }
+    }
}
```

However, this breaks the 1% edge case where you only use `ValueChanged` without `Value`. We’re just listening for changes without setting a new value for `Value`, like in this example:

https://github.com/MudBlazor/MudBlazor/blob/6d9bec89a9ee88a7dd8b85f1678e4dc4c34913e8/src/MudBlazor.UnitTests.Viewer/TestComponents/Hidden/SimpleMudHiddenTest.razor#L1-L3

In this scenario, `Hidden` becomes static because we never set a new value (either `bind-Hidden` or `<MudHidden Hidden="@_hidden" HiddenChanged="@( x => _hidden = x )" >`. Currently, I don’t know of a good way to detect if `Hidden` is not being used at all.

However, this is less problematic than the current logic in the PR I described here: https://github.com/MudBlazor/MudBlazor/pull/10544#issuecomment-2566749758.

So, I would recommend going with the `HasDelegate` check, but keep in mind that it still breaks this 1% edge case.

ScarletKuro commented:
> waiting a millisecond

`await Task.Yield()` is always better because the whole point is to have a truly async method that will pause `OnValueChanged`, allowing the Blazor lifecycle to proceed. Once the lifecycle is complete, it will continue `OnValueChanged`. It doesn't introduce additional delay, but it is still noticeable on the UI when the value switches.
With @versile2 solution is just the interception is invisible on the UI.

henon commented:
Right, I remembered about the other issue (#9507) and rediscovered the trick with `Task.Yield` and updated my snippet to also include the anti-flicker trick I found in that issue.  https://try.mudblazor.com/snippet/wuQfEFavovUKiSSV

> So, I would recommend going with the HasDelegate check, but keep in mind that it still breaks this 1% edge case.

Are you sure? But if this PR breaks the edge case and we are fine with that (are we??) then we could also do this directly in parameter state. 

I personally think that we do not need to make any changes, there is a usable workaround after all. But my position is not rigid. If you can argument why it is better to make a change then let's do it.

ScarletKuro commented:
> Are you sure?

Only if we want to make that kind of change at the toggle component level to "fix" the issue. Since the `HasDelegate` check is better than checking if the `Value is null`, as the former is a much more common scenario:

```razor
<MudToggleGroup ... Value="@("No")">
    ...
    <MudToggleItem Value="@("No")" Text="No" />
</MudToggleGroup>
```

versus

```razor
<MudToggleGroup ... ValueChanged="@(x => HiddenChangedHistory.Add(x))">
    ...
</MudToggleGroup>
```

I expect the latter to be very rare. In fact, even within our entire codebase, we only have one instance of such usage.

If we don't want to support that at all https://github.com/MudBlazor/MudBlazor/issues/10535 and offer the workaround that you posted, we could simply remove this logic from the PR. I believe the refactor itself in the PR is good, as I like how `OnAfterRender` and `OnValueChanged` are using a single `SetItemsSelected` now as code looks more simplified, what you think @henon?

henon commented:
> So, I would recommend going with the `HasDelegate` check, but keep in mind that it still breaks this 1% edge case.

Hmm, on second thought, I think you are right. We can choose between having to answer issues like this or answer issues like that. The latter is probably less often ;). So why not change it directly in ParameterState then?

henon commented:
> I believe the refactor itself in the PR is good, as I like how `OnAfterRender` and `OnValueChanged` are using a single `SetItemsSelected`, now as code looks more simplified, what you think @henon?

Sure

ScarletKuro commented:
> So why not change it directly in ParameterState then?

I was thinking about that, as it would fix that too https://github.com/MudBlazor/MudBlazor/issues/10440
Was just kind of delaying that as I wasn't sure what would be the best thing.

henon commented:
Okay, I think we can declare usage of the `...Changed` handler without a two-way binding illegal in MudBlazor. We will probably find a good spot in the docs to notify people about that rule.

versile2 commented:
I'm out of pocket today but tomorrow I can make the change. I just want to clarify, previously it just setitemselected repeatedly and this just says hey we are going to do this in the on changed event except if the user doesn't use that event then we'll do it here.i can set it to just run the method and extra time at a very low cost or add the logic that says I know you are going to do it so I won't unless you aren't.

versile2 committed:
adjusted

versile2 commented:
Now I have been out of pocket most of this week, however this should still fix the issue listed since the order changed. One of the previous method things was Deselecting values more than once in the cycle and running the Selection on init, toggle, and value(s)changed. I just refactored it to one method for init/toggle and removed the duplicate deselection logic and the hasdelegate just says hey I'm going to run the selection logic onvaluechanged let's not run it twice.

henon commented:
Thanks Versile!

ScarletKuro commented:
Just a quick question. When `ParameterState` will support interception as discussed above, it is as simple as just removing this line:
```diff
protected internal async Task ToggleItemAsync(MudToggleItem<T> item)
{
	var itemValue = item.Value;

-	var isValueBound = ValueChanged.HasDelegate;
-	var isSelectedValuesBound = ValuesChanged.HasDelegate;

	if (SelectionMode == SelectionMode.MultiSelection)
	{
		var selectedValues = new HashSet<T?>(_values.Value ?? Array.Empty<T?>());

		if (!selectedValues.Remove(itemValue))
		{
			selectedValues.Add(itemValue);
		}

		await _values.SetValueAsync(selectedValues);
-		if (isSelectedValuesBound)
-		{
-			// if SelectedValuesBound we don't need to run this method twice
-			SetItemsSelected();
-		}
	}
	else if (SelectionMode == SelectionMode.ToggleSelection)
	{
		if (EqualityComparer<T>.Default.Equals(_value.Value, itemValue))
		{
			await _value.SetValueAsync(default);
		}
		else
		{
			await _value.SetValueAsync(itemValue);
		}
-		if (isValueBound)
-		{
-			// if ValueBound we don't need to run this method twice
-			SetItemsSelected();
-		}
	}
	else // SingleSelection
	{
		await _value.SetValueAsync(itemValue);
-		if (isValueBound)
-		{
-			// if ValueBound we don't need to run this method twice
-			SetItemsSelected();
-		}
	}
}
```

versile2 commented:
> ```diff
> await _values.SetValueAsync(selectedValues);
> ```

yes if `await _values.SetValueAsync(selectedValues);` ends up ensuring the ValueChanged (or ValusChanged) event fires there's no need to fire it inside Toggle. The old behavior just did it twice just in case.
